### PR TITLE
Add proxy support for futures and spot clients (async and sync)

### PR DESCRIPTION
**Pull request by Graeme22**
# Summary
For context see: #256
This PR adds support for HTTP proxies to use with spot and futures REST clients, both sync and async.
For now websockets are omitted as the `websockets` library does not support proxies. The motivation is to allow for privacy/unblocking or other security-related concerns.

## Changes / Features
A single argument, `proxy`, has been added to the constructors for all REST clients. This proxy is used for all requests placed with those clients. Tests have also been added for all four variations: `SpotClient`, `SpotAsyncClient`, `FuturesClient`, and `FuturesAsyncClient`. The most relevant changes are included in the `kraken/base_api/__init__.py` file and in the new tests.

Closes: #256

### CODE REVIEW

File: tests/futures/test_futures_base_api.py [hunk 1]
```diff
@@ -6,8 +6,10 @@
 """Module that checks the general Futures Base API class."""
 
 from asyncio import run
+from unittest import IsolatedAsyncioTestCase
 
 import pytest
+from proxy import TestCase
```
Comment (btschwertfeger) [inline tests/futures/test_futures_base_api.py:12]: Is that the usage of the new dependency "proxy.py"?

File: tests/futures/test_futures_base_api.py [hunk 2]
```diff
@@ -119,3 +121,37 @@ async def check() -> None:
             await client.async_close()
 
     run(check())
+
+
+class TestProxyPyEmbedded(TestCase, IsolatedAsyncioTestCase):
+    def get_proxy_str(self) -> str:
+        return f"http://127.0.0.1:{self.PROXY.flags.port}"
+
+    def test_futures_rest_proxies(self) -> None:
```
Comment (btschwertfeger) [inline tests/futures/test_futures_base_api.py:130]:
```suggestion
    @pytest.mark.futures()
    @pytest.mark.futures_market()
    def test_futures_rest_proxies(self) -> None:
```

File: tests/futures/test_futures_base_api.py [hunk 3]
```diff
@@ -119,3 +121,37 @@ async def check() -> None:
...
+                "GET",
+                "/api/charts/v1/spot/PI_XBTUSD/1h",
+                auth=False,
+                post_params={"from": "1668989233", "to": "1668999233"},
+            ),
+            dict,
+        )
+
+    @pytest.mark.asyncio()
+    async def test_futures_rest_proxies_async(self) -> None:
```
Comment (btschwertfeger) [inline tests/futures/test_futures_base_api.py:146]:
```suggestion
    @pytest.mark.futures()
    @pytest.mark.futures_market()
    async def test_futures_rest_proxies_async(self) -> None:
```

File: tests/spot/test_spot_base_api.py [hunk 4]
```diff
@@ -225,3 +227,36 @@ async def check() -> None:
             await client.async_close()
 
     run(check())
+
+
+class TestProxyPyEmbedded(TestCase, IsolatedAsyncioTestCase):
+    def get_proxy_str(self) -> str:
+        return f"http://127.0.0.1:{self.PROXY.flags.port}"
+
+    def test_spot_rest_proxies(self) -> None:
```
Comment (btschwertfeger) [inline tests/spot/test_spot_base_api.py:236]:
```suggestion
    @pytest.mark.spot()
    @pytest.mark.spot_market()
    def test_spot_rest_proxies(self) -> None:
```

File: tests/spot/test_spot_base_api.py [hunk 5]
```diff
@@ -225,3 +227,36 @@ async def check() -> None:
...
+            client.request(
+                "GET",
+                "/0/public/OHLC",
+                params={"pair": "XBTUSD"},
+                auth=False,
+            ),
+        )
+
+    @pytest.mark.asyncio()
+    async def test_spot_rest_proxies_async(self) -> None:
```
Comment (btschwertfeger) [inline tests/spot/test_spot_base_api.py:251]:
```suggestion
    @pytest.mark.spot()
    @pytest.mark.spot_market()
    @pytest.mark.asyncio()
    async def test_spot_rest_proxies_async(self) -> None:
```

**Review by btschwertfeger — CHANGES_REQUESTED** [review]
Hey @Graeme22, after applying the suggested changes, I'm happy with the proposed change.

I see that the CI jobs didn't start, have you activated those on your feature branch? Locally they just ran through, but having those demonstrated here as well would be preferable.

File: tests/futures/test_futures_base_api.py (hunk 1 above)
Comment (Graeme22) [reply tests/futures/test_futures_base_api.py:12]: Correct. This is a class that extends `unittest.TestCase` that automatically sets up and tears down the proxy.

**Comment by Graeme22** [comment]
> Hey @Graeme22, after applying the suggested changes, I'm happy with the proposed change.

Great!

> I see that the CI jobs didn't start, have you activated those on your feature branch? Locally they just ran through, but having those demonstrated here as well would be preferable.

I just activated them, looks like you need to approve the changes for them to run.

**Review by btschwertfeger — APPROVED** [review]

**Comment by btschwertfeger** [comment]
> I just activated them, looks like you need to approve the changes for them to run.

I can't see anything to approve for the actions. I'll check that behavior using another account tomorrow.

**Comment by btschwertfeger** [comment]
Merging by hand now - with failing tests - since they're just failing due to wrong setup which is not that important for this change. (Some environment variables are not set in the author's repository.)

**Pull request merged**